testing/base.py

```diff
@@ -0,0 +1,73 @@
+import os
+import shutil
+import unittest
+
+from six import with_metaclass
+
+import pytest
+
+
+TEST_DOC_DIR = os.path.join(os.path.dirname(__file__), 'doc')
+
+
+def make_default_config(project):
+    """
+    Creates a default configuration for exhale
+    """
+    return {
+        'breathe_projects': {
+            project: './_doxy/xml'
+        },
+        'breathe_default_project': project,
+        'exhale_args': {
+            'containmentFolder': './api',
+            'rootFileName': 'index.rst',
+            'rootFileTitle': 'API Documentation',
+            'doxygenStripFromPath': '..',
+            'exhaleExecutesDoxygen': True,
+            'exhaleDoxygenStdin': 'INPUT = ../projects/%s/include' % project
+        }
+    }
+
+
+class ExhaleTestCaseMetaclass(type):
+    """
+    Metaclass to enforce mandatory attributes on ExhaleTestCase
+    """
+
+    def __new__(mcs, name, bases, attrs):
+        if attrs['__module__'] != __name__ and attrs.get('test_project', None) is None:
+            raise RuntimeError('ExhaleTestCase subclasses must define a "test_project" attribute')
+
+        has_tests = False
+
+        for name, attr in attrs.items():
+            if callable(attr) and name.startswith('test_'):
+                attrs[name] = pytest.mark.sphinx(
+                    testroot=TEST_DOC_DIR,
+                    confoverrides=attrs.get('config', None) or make_default_config(attrs['test_project'])
+                )(attr)
+                has_tests = True
+
+        if has_tests:
+            def _set_app(self, app):
+                self.app = app
+                yield
+                # cleanup
+                dirs = [app.config.exhale_args['containmentFolder'], os.path.dirname(app.outdir)]
+                for d in app.config.breathe_projects.values():
+                    # get only first directory of breathe project directory relative to srcdir
+                    dirs.append(os.path.relpath(os.path.join(app.srcdir, d), app.srcdir).split(os.sep)[0])
+                for d in dirs:
+                    if not os.path.isabs(d):
+                        d = os.path.join(app.srcdir, d)
+                    shutil.rmtree(d)
+            attrs['_set_app'] = pytest.fixture(autouse=True)(_set_app)
```

Comment:
Yep, you got it right.

_set_app is actually a py.test fixture (I'm using the pytext.fixture decorator, as a function). With ``autouse=True``, I tell py.test to run this fixture for every test in this class. It is getting the ``app`` in the first place - and this is where py.test's magic hits - simply because it has an argument named ``app`` and there is an existing fixture named ``app`` (in ``sphinx.testing.fixtures``). And after the ``yield`` statement the test is over and we cleanup the mess we've done :).

If you define ``test_foo`` and ``test_bar`` methods, the sphinx project will indeed be launched and initialized twice, with the exact same configuration options (I'd want to change that, probably using a basic custom decorator and/or detect a previous use of @pytest.mark.sphinx in the metaclass, but for the moment the aim was to exactly understand and highlight how it worked!).

My only concern with this pytest.fixture approach for the moment is we can't just run the configuration stage (i.e. just call ``configs.apply_sphinx_configurations``) to test how exhale extracts its configuration without generating the rst files on a project. But I'm quite sure it'll be possible to do that with a bit more digging.

Thanks to your efforts it's definitely cleaner now than my first approach with ``configure`` and ``run_exhale`` though!

Reply:
> simply because it has an argument named `app` and there is an existing fixture named `app` (in `sphinx.testing.fixtures`)

Woah.  That's incredible!

> he sphinx project will indeed be launched and initialized twice, with the exact same configuration options

I view this as less of an issue actually.  We can simply restrict names that get decorated in the metaclass.  For example, `test_after_full_execution` gets decorated the way you are doing now.  For tests that want to inspect the exhale configurations (since currently by the time the test method executes exhale is finished), we could actually cheat and enable an exhale-specific internal setting.  E.g. if the base class defines `test_after_apply_sphinx_configurations`, then the decorated `confoverrides` could set something like `exhale_args["TEST STOP AFTER"] = "apply_sphinx_configurations"`.  This would not be very difficult for me to implement, and also serves as a nice debugging feature if people really want it.

AKA for the time being, I'll just decorate `test_after_full_execution` to have it ready for windows, and then in the distance expose better "break points" in the execution lifecycle that can be tested individually.

I think this will work well, since after the `yield` the cleanup happens removing any generated files, right?

Reply:
> We can simply restrict names that get decorated in the metaclass
Hum, I'm not really keen to do that, this introduces more magic on top of py.test's magic. Instead I've added a @no_run decorator (see below), which I find much easier to understand.

> we could actually cheat and enable an exhale-specific internal setting
I tend to think that if the app code can be structured to help testing, there should not be any specific testing code in the app itself. What I've done with the @no_run decorator below is to monkeypatch ``deploy.explode()`` so that it does nothing when required. This way there is no need to clutter the code with some testing or debugging if/else statements!

It's generally better to break down the code in functions or methods that can be tested or disabled individually than adding 'break points' in a huge function.

Reply:
;)
> Looking at the new code (@no_run implementation), I see what you mean
Yep. Of course for ``@no_run`` I just adapted to the existing codebase, so it might use a better approach than monkey-patching in the future if the code is made more test-friendly !!!